### PR TITLE
Avoid `hygiene_data` lookups

### PR DESCRIPTION
**Pull request by nnethercote**
These commits mostly introduce compound operations that allow two close adjacent `hygiene_data` lookups to be combined.

r? @petrochenkov

### CODE REVIEW

**Comment by nnethercote** [comment]
Local results were very good, let's hope the CI results are similar. Instruction counts:
```
packed-simd-check
        avg: -6.4%      min: -10.2%     max: -0.5%
helloworld-check
        avg: -2.4%      min: -2.8%      max: -2.2%
unify-linearly-check
        avg: -1.9%      min: -2.7%      max: -1.3%
deeply-nested-check
        avg: -1.6%      min: -2.7%      max: -1.1%
script-servo-check
        avg: -2.0%      min: -2.6%      max: -1.6%
webrender-check
        avg: -1.8%      min: -2.4%      max: -1.2%
cranelift-codegen-check
        avg: -1.8%      min: -2.3%      max: -1.4%
style-servo-check
        avg: -1.8%      min: -2.1%      max: -1.5%
issue-46449-check
        avg: -1.5%      min: -1.9%      max: -1.3%
inflate-check
        avg: -0.5%      min: -1.6%      max: -0.1%
crates.io-check
        avg: -1.1%      min: -1.5%      max: -0.6%
regression-31157-check
        avg: -1.0%      min: -1.5%      max: -0.6%
sentry-cli-check
        avg: -1.2%      min: -1.5%      max: -0.9%
regex-check
        avg: -1.1%      min: -1.4%      max: -1.0%
tokio-webpush-simple-check
        avg: -1.0%      min: -1.4%      max: -0.7%
ripgrep-check
        avg: -1.0%      min: -1.2%      max: -0.8%
wg-grammar-check
        avg: -0.6%      min: -1.2%      max: -0.2%
futures-check
        avg: -0.7%      min: -1.1%      max: -0.4%
serde-check
        avg: -0.7%      min: -1.0%      max: -0.5%
piston-image-check
        avg: -0.8%      min: -1.0%      max: -0.6%
encoding-check
        avg: -0.8%      min: -1.0%      max: -0.5%
syn-check
        avg: -0.7%      min: -1.0%      max: -0.5%
clap-rs-check
        avg: -0.6%      min: -0.9%      max: -0.5%
hyper-check
        avg: -0.7%      min: -0.9%      max: -0.5%
unused-warnings-check 
        avg: -0.6%      min: -0.8%      max: -0.5%
```

@bors try

**Comment by nnethercote** [comment]
@Manishearth: the new `outer_expn_info()` function introduced in the second commit can be used in Clippy -- every single occurence of `outer()` in Clippy is immediately followed by `expn_info()`. (Note that some of these occurrences span two lines, and won't be found with a simple grep.) It's possible that using the new function might speed up Clippy.

**Comment by Manishearth** [comment]
Thanks for the heads up!

**Comment by nnethercote** [comment]
@rust-timer build 9f4ea3d6210445892cbd666b2413dcf95a68a363

**Comment by nnethercote** [comment]
@Mark-Simulacrum: rust-timer went nuts there for a bit. Any idea what might have gone wrong?

**Comment by nnethercote** [comment]
The perf results on CI are good, similar to what I saw locally.

**Comment by michaelwoerister** [comment]
This looks pretty good, @nnethercote!

**Comment by Mark-Simulacrum** [comment]
Will investigate, but definitely highly unexpected behavior. Odd. Can't think what could've caused it.

**Comment by nnethercote** [comment]
@Mark-Simulacrum: In case it helps:
- It added one comment per minute between 7:00pm and 7:30pm (GMT+10).
- While the comments were being added, the results weren't available at the comparison URL -- it showed that "commit not found" error that sometimes occurs.
- Once the results became available at the comparison URL, the comments stopped, as far as I could tell.

**Comment by Mark-Simulacrum** [comment]
My working theory is that we were going over memory budget and getting killed and restarted, just to realize that we're done (ping from server) and then die right before we could save.

**Comment by petrochenkov** [comment]
Suggestion: make `outer()`, `expn_info()` and `is_descendant_of()` private methods on `HygieneData` and call them from hte `Mark`/`SyntaxCtxt` methods.

(Everything else looks good, but I didn't look at the last commit yet, will look later.)

File: src/librustc/ty/mod.rs [hunk 1]
```diff
@@ -3089,17 +3089,27 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
...
+        ident
+    }
+
+    pub fn adjust_ident_and_get_scope(self, mut ident: Ident, scope: DefId, block: hir::HirId)
+                                      -> (Ident, DefId) {
+        ident = ident.modern();
+        let scope = match ident.span.adjust(self.target_expansion(scope)) {
             Some(actual_expansion) =>
                 self.hir().definitions().parent_module_of_macro_def(actual_expansion),
             None if block == hir::DUMMY_HIR_ID => DefId::local(CRATE_DEF_INDEX), // Dummy DefId
```
Comment (petrochenkov) [inline src/librustc/ty/mod.rs]: This case was introduced specifically for the `DUMMY_HIR_ID` argument in `fn hygienic_eq`, so it should be dead code now.

File: src/librustc/ty/mod.rs [hunk 2]
```diff
@@ -3085,16 +3085,31 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
...
+        // `modern()` calls required for the span comparison whenever possible.
+        use_name.name == def_name.name &&
+        self.adjust_ident(use_name, def_parent_def_id).span.ctxt() == def_name.modern().span.ctxt()
     }
 
-    pub fn adjust_ident(self, mut ident: Ident, scope: DefId, block: hir::HirId) -> (Ident, DefId) {
-        ident = ident.modern();
-        let target_expansion = match scope.krate {
+    #[inline]
+    fn target_expansion(self, scope: DefId) -> Mark {
```
Comment (petrochenkov) [inline src/librustc/ty/mod.rs]:
```suggestion
    fn expansion_that_defined(self, scope: DefId) -> Mark {
```
(`let target_expansion` wasn't an example of good variable naming.)

Also, inline is usually unnecessary on private functions?

**Comment by nnethercote** [comment]
@petrochenkov: New code is up. I have addressed your comments. The new functions are in a new commit at the end of the series.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Comment by Centril** [comment]
@bors rollup=never 

for perf.

**Comment by Centril** [comment]
@bors p=1 

because never.

**Comment by nnethercote** [comment]
Final perf improvements are [here](https://perf.rust-lang.org/compare.html?start=aee7012fab26d5e307a2fe767e4e7c847c5a45ee&end=0bfbaa6e8dfb509b453020740fd37c7a22882c87&stat=instructions:u).

**Pull request merged**